odoo/addons/base/models/assetsbundle.py

```diff
@@ -322,13 +342,71 @@ def save_attachment(self, type, content):
 
         return attachment
 
-    def js(self):
-        attachments = self.get_attachments('js')
-        if not attachments:
-            content = ';\n'.join(asset.minify() for asset in self.javascripts)
-            return self.save_attachment('js', content)
+    def js(self, is_minified=True, with_sourcemap=False, name_unique=True):
+        url = None if name_unique else "/%s.js" % self.name
+        attachments = self.get_attachments('js', is_minified=is_minified, url_pattern=url)
+        if not attachments or (not name_unique and bundle_checksum.get(self.name, None) != self.checksum):
+            if not name_unique:
+                bundle_checksum[self.name] = self.checksum
+            if (with_sourcemap):
+                return self.js_with_sourcemap(name_unique=name_unique)
+            else:
+                content = ';\n'.join(asset.minify() if is_minified else asset.with_header(asset.content, minimal=False)
+                                    for asset in self.javascripts)
+                return self.save_attachment('js', content, is_minified=is_minified, url=url)
+
         return attachments[0]
 
+    def js_with_sourcemap(self, name_unique=True):
+        content_bundle = ""
+        generator = SourceMapGenerator({
+            "file": "/%s.js" % self.name,
+            "sourceRoot": "./"
+        })
+        content_line_count = 0
+        line_header = 6
+        for asset in self.javascripts:
+            if asset.is_odoo_module:
+                generator.add_source(
+                    asset.url, asset._content, content_line_count, nb_skip_line_output=line_header+3)
+                asset_line_count = len(asset.content.split("\n")) + 6
+            else:
+                asset_line_count = generator.add_source(
+                    asset.url, asset.content, content_line_count, nb_skip_line_output=line_header)
+
+            if (content_bundle == ""):
+                content_bundle = asset.with_header(asset.content, minimal=False)
+            else:
+                content_bundle += ';\n' + asset.with_header(asset.content, minimal=False)
```

Comment:
`content_bundle` should be a list, these should just append to the list, and once we actually want a string the list shoudl be joined using `';\n'`.